lib/cn/node_split.c

```diff
@@ -458,3 +462,79 @@ cn_split(struct cn_compaction_work *w)
 
     return 0;
 }
+
+static bool
+check_valid_kvsets(const struct cn_compaction_work *w, bool left)
+{
+    uint start, end;
+
+    if (left) {
+        start = 0;
+        end = w->cw_kvset_cnt;
+    } else {
+        start = w->cw_kvset_cnt;
+        end = w->cw_outc;
+    }
+
+    for (uint i = start; i < end; i++) {
+        if (w->cw_kvsetidv[i] != CNDB_INVAL_KVSETID)
+            return true;
+    }
+
+    return false;
+}
+
+merr_t
+cn_split_nodes_alloc(
+    const struct cn_compaction_work *w,
+    uint64_t                         nodeidv[static 2],
+    struct cn_tree_node             *nodev[static 2])
+{
+    for (int i = LEFT; i <= RIGHT; i++) {
+        nodev[i] = NULL;
+
+        if (!check_valid_kvsets(w, i == LEFT))
+            continue;
+
+        if (i == LEFT) {
+            struct cn_tree *tree = w->cw_tree;
+
+            nodeidv[i] = cndb_nodeid_mint(cn_tree_get_cndb(w->cw_tree));
+
+            nodev[i] = cn_node_alloc(tree, nodeidv[i]);
+            if (!nodev[i])
+                return merr(ENOMEM);
+
+            /* Allocate a route node using the split key as its edge key.
+             */
+            nodev[i]->tn_route_node =
+                route_node_alloc(tree->ct_route_map, nodev[i], w->cw_split.key, w->cw_split.klen);
+            if (!nodev[i]->tn_route_node) {
+                cn_node_free(nodev[i]);
+                nodev[i] = NULL;
+
+                return merr(ENOMEM);
+            }
+        } else {
+            /* Use the source node as the right node */
+            nodeidv[i] = w->cw_node->tn_nodeid;
+            nodev[i] = w->cw_node;
+        }
+    }
```

Comment:
Seems like it'd be more straightforward if you got rid of the loop...

Reply:
You're right. Now it looks cleaner, thanks!